src/transform-sdk/cpp/src/transform_sdk.cc

```diff
@@ -0,0 +1,632 @@
+
+/**
+ * A simple abort function that prints then aborts
+ */
+[[noreturn]] void abort(std::string_view str) {
+    println(str);
+    std::abort();
+}
+
+namespace abi {
+#ifdef __wasi__
+
+extern "C" {
+#define WASM_IMPORT(mod, name)                                                 \
+    __attribute__((import_module(#mod), import_name(#name)))
+
+WASM_IMPORT(redpanda_transform, check_abi_version_1)
+void redpanda_transform_check();
+constexpr auto check = redpanda_transform_check;
+
+WASM_IMPORT(redpanda_transform, read_batch_header)
+int32_t redpanda_transform_read_batch_header(
+  int64_t* base_offset,
+  int32_t* record_count,
+  int32_t* partition_leader_epoch,
+  int16_t* attributes,
+  int32_t* last_offset_delta,
+  int64_t* base_timestamp,
+  int64_t* max_timestamp,
+  int64_t* producer_id,
+  int16_t* producer_epoch,
+  int32_t* base_sequence);
+constexpr auto read_batch_header = redpanda_transform_read_batch_header;
+
+WASM_IMPORT(redpanda_transform, read_next_record)
+int32_t redpanda_transform_read_next_record(
+  uint8_t* attributes,
+  int64_t* timestamp,
+  int64_t* offset,
+  uint8_t* buf,
+  uint32_t len);
+constexpr auto read_next_record = redpanda_transform_read_next_record;
+
+WASM_IMPORT(redpanda_transform, write_record)
+int32_t redpanda_transform_write_record(uint8_t* buf, uint32_t len);
+constexpr auto write_record = redpanda_transform_write_record;
+
+#undef WASM_IMPORT
+}
+
+#else
+
+void check() { abort("check_abi - stub"); }
+
+int32_t read_batch_header(
+  int64_t* /*unused*/,
+  int32_t* /*unused*/,
+  int32_t* /*unused*/,
+  int16_t* /*unused*/,
+  int32_t* /*unused*/,
+  int64_t* /*unused*/,
+  int64_t* /*unused*/,
+  int64_t* /*unused*/,
+  int16_t* /*unused*/,
+  int32_t* /*unused*/) {
+    abort("read_batch_header - stub");
+}
+
+int32_t read_next_record(
+  uint8_t* /*unused*/,
+  int64_t* /*unused*/,
+  int64_t* /*unused*/,
+  uint8_t* /*unused*/,
+  uint32_t /*unused*/) {
+    abort("read_next_record - stub");
+}
+
+int32_t write_record(const uint8_t* /*unused*/, uint32_t /*unused*/) {
+    abort("write_record - stub");
+}
+
+#endif
+} // namespace abi
+
+namespace varint {
+
+constexpr size_t MAX_LENGTH = 10;
+
+template<typename T>
+struct decoded {
+    T value;
+    size_t read;
+};
+
+uint64_t zigzag_encode(int64_t num) {
+    constexpr unsigned int signbit_shift = (sizeof(int64_t) * CHAR_BIT) - 1;
+    // NOLINTNEXTLINE
+    return (static_cast<uint64_t>(num) << 1) ^ (num >> signbit_shift);
+}
+
+int64_t zigzag_decode(uint64_t num) {
+    // NOLINTNEXTLINE
+    return static_cast<int64_t>(num >> 1) ^ -static_cast<int64_t>(num & 1);
+}
+
+std::expected<decoded<uint64_t>, std::error_code>
+read_unsigned(bytes_view payload) {
+    uint64_t value = 0;
+    unsigned int shift = 0;
+    for (size_t i = 0; i < payload.size(); ++i) {
+        const uint8_t byte = payload[i];
+        if (i >= MAX_LENGTH) {
+            return std::unexpected(
+              std::make_error_code(std::errc::value_too_large));
```

Comment:
it's so beautiful 🥺 😭

Reply:
I know I do love `-fno-exceptions`. I guess it's how I learned to write C++.